Libraries/Text/RCTText.m

```diff
@@ -149,6 +165,101 @@ - (void)didMoveToWindow
+
+  __block BOOL hitMinimumScale = NO;
+  while ((requiredSize.height > CGRectGetHeight(frame) ||
+         requiredSize.width > CGRectGetWidth(frame) ||
+         (linesRequired > textContainer.maximumNumberOfLines &&
+          textContainer.maximumNumberOfLines != 0))
+         && !hitMinimumScale)
+  {
+    [_textStorage beginEditing];
+    [_textStorage enumerateAttribute:NSFontAttributeName
+                             inRange:glyphRange
+                             options:0
+                          usingBlock:^(UIFont *font, NSRange range, BOOL *stop)
+    {
+      if (font) {
+        UIFont *originalFont = [_originalString attribute:NSFontAttributeName
+                                                  atIndex:range.location
+                                           effectiveRange:&range];
+        UIFont *newFont = [font fontWithSize:font.pointSize - .5];
+        if (newFont.pointSize > originalFont.pointSize * self.minimumFontScale) {
+          [_textStorage removeAttribute:NSFontAttributeName range:range];
+          [_textStorage addAttribute:NSFontAttributeName value:newFont range:range];
+        } else {
+          hitMinimumScale = YES;
+        }
+      }
+    }];
+    [_textStorage endEditing];
+
+    linesRequired = [self numberOfLinesRequired:layoutManager];
+    requiredSize = [self calculateSize:_textStorage];
+  }
+
+  //Vertically center draw position
+  frame.origin.y = _contentInset.top + round((CGRectGetHeight(frame) - requiredSize.height) / 2);
+  return frame;
+}
+
+// Via Apple Text Layout Programming Guide
+// https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/TextLayout/Tasks/CountLines.html
+- (NSInteger)numberOfLinesRequired:(NSLayoutManager *)layoutManager
+{
+  NSInteger numberOfLines, index, numberOfGlyphs = [layoutManager numberOfGlyphs];
+  NSRange lineRange;
+  for (numberOfLines = 0, index = 0; index < numberOfGlyphs; numberOfLines++){
+    (void) [layoutManager lineFragmentRectForGlyphAtIndex:index
+                                           effectiveRange:&lineRange];
+    index = NSMaxRange(lineRange);
+  }
+
+  return numberOfLines;
+}
+
+// Via Apple Text Layout Programming Guide
+//https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/TextLayout/Tasks/StringHeight.html
+- (CGSize)calculateSize:(NSTextStorage *)storage
+{
+  NSLayoutManager *layoutManager = [storage.layoutManagers firstObject];
+  NSTextContainer *textContainer = [layoutManager.textContainers firstObject];
+  (void) [layoutManager glyphRangeForTextContainer:textContainer];
+  return [layoutManager usedRectForTextContainer:textContainer].size;
+}
+
+//Start fresh with the original drawn string each time, in case frame has gotten larger
+- (void)resetDrawnTextStorage
+{
+  [_textStorage beginEditing];
+
+  NSRange originalRange = NSMakeRange(0, _originalString.length);
+  [_textStorage setAttributes:@{} range:originalRange];
```

Comment:
So NSTextStorage is a subclass of NSAttributedString, we're purposefully fully wiping it here and replacing it with the original attributes. This way we start from scratch each time we want to autosize the text.